### PR TITLE
fix: Checks for null

### PR DESCRIPTION
**Pull request by Egor18**
Fix invalid checks for null (dereferenced before checked) (#2483 Warnings 5)

### CODE REVIEW

File: src/main/java/spoon/support/compiler/jdt/ReferenceBuilder.java [hunk 1]
```diff
@@ -307,9 +307,12 @@ private boolean isTypeArgumentExplicit(TypeReference[][] typeArguments) {
 	 */
 	CtReference getDeclaringReferenceFromImports(char[] expectedName) {
 		CompilationUnitDeclaration cuDeclaration = this.jdtTreeBuilder.getContextBuilder().compilationunitdeclaration;
+		if (cuDeclaration == null) {
```
Comment (monperrus) [inline src/main/java/spoon/support/compiler/jdt/ReferenceBuilder.java]: what's the added value of this change?

File: src/main/java/spoon/support/reflect/declaration/CtEnumImpl.java [hunk 2]
```diff
@@ -113,8 +113,12 @@ public boolean removeEnumValue(CtEnumValue<?> enumValue) {
 
 	@Override
 	public <C extends CtEnum<T>> C setEnumValues(List<CtEnumValue<?>> enumValues) {
+		if (enumValues == null) {
```
Comment (monperrus) [inline src/main/java/spoon/support/reflect/declaration/CtEnumImpl.java]: Why does this remove the static analysis warning?

File: src/main/java/spoon/pattern/PatternBuilder.java [hunk 3]
```diff
@@ -112,11 +112,11 @@ public static PatternBuilder create(CtElement... elems) {
 	}
 
 	protected PatternBuilder(List<CtElement> template) {
-		this.templateTypeRef = getDeclaringTypeRef(template);
-		this.patternModel = Collections.unmodifiableList(new ArrayList<>(template));
 		if (template == null) {
 			throw new SpoonException("Cannot create a Pattern from an null model");
 		}
+		this.templateTypeRef = getDeclaringTypeRef(template);
```
Comment (monperrus) [inline src/main/java/spoon/pattern/PatternBuilder.java]: this is easily mergeable, what about putting it in a separate PR? 

(for changes that look quite different like this one, the discussion and merge is easier in separate PRs, thanks!)

File: src/main/java/spoon/support/compiler/jdt/ReferenceBuilder.java (hunk 1 above)
Comment (Egor18) [reply src/main/java/spoon/support/compiler/jdt/ReferenceBuilder.java]: The previous check `cuDeclaration != null` makes no sense. It was performed too late:

```
LookupEnvironment environment = cuDeclaration.scope.environment; // <= cuDeclaration dereference (if cuDeclaration is null, we will get NullPointerException)
if (cuDeclaration != null && cuDeclaration.imports != null) { ... } // <= cuDeclaration could not be null here
```

Alternatively, I can remove that check at all, if `this.jdtTreeBuilder.getContextBuilder().compilationunitdeclaration` could not be null.

File: src/main/java/spoon/support/reflect/declaration/CtEnumImpl.java (hunk 2 above)
Comment (Egor18) [reply src/main/java/spoon/support/reflect/declaration/CtEnumImpl.java]: The same pattern here:
```
.... new ArrayList<>(enumValues) // <= if enumValues is null, we will get NullPointerException
if (enumValues == null || enumValues.isEmpty()) { .... } // <= enumValues could not be null here
```

We detect such errors with data-flow analysis. 
When I add that check for null with unconditional return in 'then' branch, the analyzer understands that in the code, following the check, the `enumValues` reference is not null.

File: src/main/java/spoon/pattern/PatternBuilder.java (hunk 3 above)
Comment (Egor18) [reply src/main/java/spoon/pattern/PatternBuilder.java]: Would you like me to make 3 separate PR's for this issue again?

File: src/main/java/spoon/support/reflect/declaration/CtEnumImpl.java (hunk 2 above)
Comment (monperrus) [reply src/main/java/spoon/support/reflect/declaration/CtEnumImpl.java]: Thanks. I was not aware that `new ArrayList<>(null)` throws an NPE

**Pull request merged**